lib/Dialect/TritonGPU/Transforms/Pipeliner/Schedule.h

```diff
@@ -10,11 +10,20 @@
 namespace mlir {
 namespace triton {
 
-/// This fill out the pipelining options including schedule and annotations for
-/// wait ops. This also does pre-processing by converting some of the loads into
-/// async loads so that the IR is ready to be pipelined.
-bool preProcessLoopAndGetSchedule(scf::ForOp &forOp, int numStages,
-                                  mlir::triton::PipeliningOption &options);
+class MatmulPipelineSchedule {
+public:
+  /// This fill out the pipelining options including schedule and annotations
+  /// for wait ops. This also does pre-processing by converting some of the
+  /// loads into async loads so that the IR is ready to be pipelined.
+  bool preProcessLoopAndGetSchedule(scf::ForOp &forOp, int numStages,
+                                    mlir::triton::PipeliningOption &options);
+
+  /// Post process the pipelined loop by inserting sync operations if necessary
+  void insertWaits(RewriterBase &rewriter, scf::ForOp forOp);
+
+private:
+  llvm::SmallVector<mlir::Operation *, 4> extractOps;
```

Comment:
Discussed this one with @ThomasRaoux and I'll replace this mechanism with marking the Ops that need to change with attributes. This whole insertWaits mechanism can then be run after the pipeliner is done with entire module. It will be more generic than building it exclusively for Matmul pipeline.

Reply:
Makes sense. It will still be as a post process of the pipeliner right ?

Reply:
Right!